contents/handbook/engineering/async-migrations.md

```diff
@@ -0,0 +1,222 @@
+---
+title: Async Migrations
+sidebar: Handbook
+showTitle: true
+---
+
+> **Note:** Async migrations are currently a work in progress under the name of "special migrations".
+
+## What are async migrations?
+
+Async migrations (AKA big/special migrations) are migrations that do not run synchronously on an update to a PostHog instance. Rather, they execute on the background of a running PostHog instance, and should be completed within a range of PostHog versions. 
+
+## Why are async migrations necessary?
+
+Migrations are inevitable, and sometimes it may be necessary to execute non-trivial schema changes that can take a long time to complete. 
+
+For example, ClickHouse does not support changing the primary key of a table, which is a change we were [forced to make in anticipation of upgrading ClickHouse beyond version 21.6](https://github.com/PostHog/posthog/issues/5684). As a result, the way to change the schema of the table was to create a new table and insert all the data from the old table into it, which took us an entire week to run on PostHog Cloud.
+
+Now, while we can execute such changes to PostHog Cloud "manually" (i.e. James G SSHs into CH5, runs a bunch of queries using clickhouse-client, and keeps an eye on them), we cannot expect the same of our users (plus this approach does not scale well for us either).
+
+As a result, we needed to create a system capable of safely and efficiently managing migrations that need to happen asynchronously.
+
+## Why did we write this system from scratch?
+
+As a provider of a product that users can self-host, we needed a bespoke solution that interoperates well with our particular tech stack and requirements, such as:
+
+1. The ability to run asynchronous tasks using a system already in use (Celery)
+2. Accounting for both databases we have in use (Postgres and ClickHouse) and their subsequent migrations
+3. Accounting for our EE codebase
+4. The ability for self-hosted users to manage and monitor migrations easily 
+5. Accounting for our particular deployment options
+6. Allowing logical branching (splitting code paths) in the future
+7. Extensibility to support future needs we might have for long-running operations
+
+## How do async migrations work?
+
+### Writing an async migration
+
+To write an async migration, you should create a migration file inside `posthog/special_migrations/migrations`. The name should follow the convention we use for Django and EE migrations (e.g. `0005_update_events_schema`).
+
+Here's what an example async migration looks like:
```

Comment:
I wrote this new one specifically for this - as it's not a nice test/example in the repo but good for educational purposes. Not sure the best approach here. People are also less likely to visit the link than read inline stuff

Reply:
true, but it will get out of date (especially since it requires PRs against 2 repos), which I believe is more important. Alternatively you can mention that there are more examples and link to them, so even if it's out of date it's not that bad.